streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -479,6 +485,26 @@ public void restore(final Map<TaskId, Task> tasks) {
         }
     }
 
+    private void updateStandbyPartitions(final Map<TaskId, Task> tasks,
+        final Set<TopicPartition> restoringChangelogs) {
+        // For standby tasks in the `tasks` map, we make sure that the `restoreConsumer`'s state
+        // is correct.
+        if (state == ChangelogReaderState.STANDBY_UPDATING) {
+            for (final TopicPartition partition : restoringChangelogs) {
+                final ProcessorStateManager manager = changelogs.get(partition).stateManager;
+                final TaskId taskId = manager.taskId();
+                final Task task = tasks.get(taskId);
+                if (manager.taskType() == TaskType.STANDBY) {
+                    if (task != null) {
+                        restoreConsumer.resume(Collections.singleton(partition));
+                    } else {
+                        restoreConsumer.pause(Collections.singleton(partition));
+                    }
```

Comment:
I think a better way would be to collect the partitions to resume and pause and then call `restoreConsumer.pause()` and `restoreConsumer.restore()` just once with the collections of partitions.

Reply:
Done in https://github.com/apache/kafka/pull/12161/commits/cb6d6f132a91c3373cbb6556d0b00b1d2bf8eb1b.